### ISSUE TITLE
Error writing snapshot backup file

### ISSUE THREAD

**Issue by niboshi** [pfnet/pytorch-pfn-extras#168]
https://github.com/pfnet/pytorch-pfn-extras/blob/1b265e7de61f94cbb0910dbd60aa86a719ebdb7d/pytorch_pfn_extras/writing.py#L216

This line caused an error.
It seems `.bak` file already existed at the time of this call. I think such a situation can easily happen if the previous training session was terminated before [removing it](https://github.com/pfnet/pytorch-pfn-extras/blob/1b265e7de61f94cbb0910dbd60aa86a719ebdb7d/pytorch_pfn_extras/writing.py#L221).

By the way, it also failed to load the snapshot of the previous session and started over from the epoch 1 (see the timestamps of the files listed below).
I don't know whether or not it's related to this issue.

```
epoch       iteration   elapsed_time  lr          train/loss  val/loss    val/top1    val/top5                                                                                                                                       
1           58          17.791        0.1         8.19181     31.9144     0           0.0119048                                                                                                                                      
2           116         40.637        0.1         7.68947     7.59687     0.00892857  0.00892857                                                                                                                                     
Traceback (most recent call last):                                                                                                                                                                                                   
  File "/usr/local/lib/python3.8/runpy.py", line 194, in _run_module_as_main                                                                                                                                                         
    return _run_code(code, main_globals, None,                                                                                                                                                                                       
  File "/usr/local/lib/python3.8/runpy.py", line 87, in _run_code                                                                                                                                                                    
    exec(code, run_globals)                                                                                                                                                                                                          
  File "mywork/__main__.py", line 40, in <module>                                      
    main()                                                                                                                                                                                                                           
  File "mywork/__main__.py", line 34, in main                                          
    args.func(args)                                                                                                                                                                                                                  
  File "mywork/train.py", line 422, in main                                            
    run_train(manager, model, device, train_loader, optimizer,                                                                                                                                                                       
  File "mywork/train.py", line 153, in run_train                                       
    ppe.reporting.report({'train/loss': loss.item()})                                                                                                                                                                                
  File "/usr/local/lib/python3.8/contextlib.py", line 120, in __exit__                                                                                                                                                               
    next(self.gen)                                                                                                                                                                                                                   
  File "/home/nishino/.local/lib/python3.8/site-packages/pytorch_pfn_extras/training/manager.py", line 430, in run_iteration                                                                                                         
    self.run_extensions()                                                                                                                                                                                                            
  File "/home/nishino/.local/lib/python3.8/site-packages/pytorch_pfn_extras/training/manager.py", line 299, in run_extensions                                                                                                        
    entry.extension(self)                                                                                                                                                                                                            
  File "/home/nishino/.local/lib/python3.8/site-packages/pytorch_pfn_extras/training/extensions/_snapshot.py", line 458, in __call__                                                                                                 
    self._make_snapshot(trainer)                                                                                                                                                                                                     
  File "/home/nishino/.local/lib/python3.8/site-packages/pytorch_pfn_extras/training/extensions/_snapshot.py", line 410, in _make_snapshot                                                                                           
    writer(filename, outdir, serialized_target, savefun=self._savefun)                                                                                                                                                               
  File "/home/nishino/.local/lib/python3.8/site-packages/pytorch_pfn_extras/writing.py", line 270, in __call__                                                             
    self.save(filename, out_dir, target, savefun, append, **self._kwds)
  File "/home/nishino/.local/lib/python3.8/site-packages/pytorch_pfn_extras/writing.py", line 213, in save                                                             
    self.fs.rename(dest, bak)                                                                                                                                                                                                        
  File "/usr/local/lib/python3.8/site-packages/pfio/filesystems/hdfs.py", line 295, in rename                                                                                                                                        
    return self.connection.rename(src, dst)                                                                                                          
  File "/usr/local/lib/python3.8/site-packages/pyarrow/hdfs.py", line 86, in rename                                                                                        
    return super().rename(path, new_path)                                           
  File "pyarrow/io-hdfs.pxi", line 178, in pyarrow.lib.HadoopFileSystem.rename                                                                                                          
  File "pyarrow/error.pxi", line 99, in pyarrow.lib.check_status                                                                                                                                                
OSError: HDFS Rename failed, errno: 5 (Input/output error)                                                                                                                              
```


```
-rw-r--r--   3 niboshi niboshi  231673368 2021-05-24 22:40 hdfs:///.../last_snapshot
-rw-r--r--   3 niboshi niboshi  231991832 2021-05-24 21:57 hdfs:///.../last_snapshot.bak
-rw-r--r--   3 niboshi niboshi       554 2021-05-24 23:14 hdfs:///.../log
-rw-r--r--   3 niboshi niboshi  115971823 2021-05-24 22:40 hdfs:///.../model_epoch_001
-rw-r--r--   3 niboshi niboshi  115971823 2021-05-24 23:14 hdfs:///.../model_epoch_002
-rw-r--r--   3 niboshi niboshi  115971823 2021-05-24 11:51 hdfs:///.../model_epoch_003
-rw-r--r--   3 niboshi niboshi  115971823 2021-05-24 11:52 hdfs:///.../model_epoch_004
...
```

**Comment by niboshi** [comment]
I encountered the same situation in another training instance.
This time, too, the training ignored the saved snapshot and started over from the epoch 1. So I'd assume it's related.
Also, the error happened just after epoch 2, just like the above instance. I don't think it's coincidence.

**Comment by emcastillo** [comment]
Thank you for the report!
I suppose you are using `autoload=True`
I suspect it may have something to do with the snapshot name format. I will try to create a reproducer.

Thank you again!

**Comment by emcastillo** [comment]
I couldn't reproduce the snapshot not loading in Unix, I suspect the issue might be related to HDFS.
this is my snapshot extension config
`extensions.snapshot(autoload=True, filename='model_epoch_{.epoch}'),` also tested 
`extensions.snapshot(autoload=True, filename='model_epoch_00{.epoch}'),`

**Comment by niboshi** [comment]
Of about 20 training instances, only two suffered this error. It might not be always reproduced in a single trial.

**Comment by dhgrs** [comment]
Here is how snapshot extensions find snapshots.

https://github.com/pfnet/pytorch-pfn-extras/blob/18cde64a6216e25f3ea0e62a700f0cbaa4b57fd9/pytorch_pfn_extras/training/extensions/_snapshot.py#L29-L33

If the filename ends with `}`, the extension only searches depending on the prefix because the suffix is empty. That's why `<filename>.bak`  can be found and you failed to reproduce.

In my env, it happens 100% on local file system when `filename="snapshot_latest"`.

**Comment by niboshi** [comment]
If `filename` is `"snapshot_latest"`, is `fmt` also `"snapshot_latest"` and so are `prefix` and `suffix`?

**Comment by emcastillo** [comment]
I think that @dhgrs refers to the case when the process has exited before being able to save the new snapshot and the old one is renamed to `snapshot_file.bak` in the file system. In this case, the autoload is not able to restart from this file depending on the `filename` format.

> If filename is "snapshot_latest", isn't fmt also "snapshot_latest" and so are prefix and suffix?

This is true for `filename="snapshot_latest"` and that is why `.bak` is not found in @dhgrs case,
but if we have `filename="snapshot_latest{}"`  then suffix is empty and the temporary file is found (wether this behavior is correct or not is a different question that we must address :))

I am not sure if this is the same problem you were facing, sounds similar but did you have valid snapshot files in the folder?

**Comment by dhgrs** [comment]
After I posted, I noticed that this issue seems slightly different from my case. It might have the same cause but I'm not sure too. 

>did you have valid snapshot files in the folder?

In my case that `filename` is `snapshot_latest`, when I found the job has restarted from the first epoch, the snapshot extension has triggered several times. There were only `snapshot_latest`. `tmp_snapshot_latest` or `snapshot_latest.bak` didn't exist. I don't know which files are in the log dir when the jobs restarted. I'll submit a new issue for my case if it's needed.
